### PR TITLE
Replace sampler arrays with constant sampler elements, simplify and reuse code for all shaders

### LINKED ISSUE #70013: Many errors and crash during iOS startup on older iPad Air
### Godot version

4.0.beta8

### System information

iPad Air 2, iOS 15.7, Apple A8X GPU

### Issue description

I've recently been testing iOS export, and it generally works very well on newer devices even though there are some errors and warnings in the console during startup. However, when trying to run my project on my slightly older iPad Air 2 I get a ton of errors and the game crashes before reaching the title screen (possibly after initial startup when it tries to load shaders and display the title screen). The title music starts playing, but then screeches to a halt and the game crashes before the splash screen is visually replaced by the title screen.

You can find the full error log here: [4.0.0beta8_ipad_air2_vulcan_errors.txt](https://www.lostminds.com/godot/4.0.0beta8_ipad_air2_vulcan_errors.txt)

We start out with this startup and GPU description:

```
2022-12-13 10:30:26.876756+0100 vlobs[378:6102] Godot Engine v4.0.beta8.official.c6e40e1c0 - https://godotengine.org
[mvk-info] MoltenVK version 1.2.0, supporting Vulkan version 1.2.231.
```
...

```
[mvk-info] GPU device:
		model: Apple A8X GPU
		type: Integrated
		vendorID: 0x106b
		deviceID: 0xf0603ea
		pipelineCacheUUID: 000027D8-0F06-03EA-0000-000000000000
	supports the following Metal Versions, GPU's and Feature Sets:
		Metal Shading Language 2.4
		GPU Family Apple 2
		GPU Family Apple 1
		GPU Family Common 1
		iOS GPU Family 2 v5
		iOS GPU Family 2 v4
		iOS GPU Family 2 v3
		iOS GPU Family 2 v2
		iOS GPU Family 2 v1
		iOS GPU Family 1 v5
		iOS GPU Family 1 v4
		iOS GPU Family 1 v3
		iOS GPU Family 1 v2
		iOS GPU Family 1 v1
[mvk-info] Created VkInstance for Vulkan version 1.2.0, as requested by app, with the following 3 Vulkan extensions enabled:
		VK_KHR_get_physical_device_properties2 v2
		VK_KHR_surface v25
		VK_MVK_ios_surface v3
```

This is then followed by a lot of errors being repeated, but the first one that repeats many times is 
```
[mvk-error] VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8X GPU does not support arrays of samplers.
[mvk-error] VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8X GPU does not support arrays of samplers.
[mvk-error] VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8X GPU does not support arrays of samplers.
[mvk-error] VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8X GPU does not support arrays of samplers.
[mvk-error] VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8X GPU does not support arrays of samplers.
2022-12-13 10:30:41.353740+0100 vlobs[378:6347] USER ERROR: Error (-8) creating descriptor set layout for set 0
2022-12-13 10:30:41.354698+0100 vlobs[378:6347]    at: shader_create_from_bytecode (drivers/vulkan/rendering_device_vulkan.cpp:5591)
```

Followed later by many repetitions of 
```
2022-12-13 10:30:46.880119+0100 vlobs[378:6102] USER ERROR: Condition "!src_texture" is true. Returning: RID()
2022-12-13 10:30:46.880334+0100 vlobs[378:6102]    at: texture_create_shared_from_slice (drivers/vulkan/rendering_device_vulkan.cpp:2292)
2022-12-13 10:30:46.880539+0100 vlobs[378:6102] USER ERROR: Attempted to name invalid ID: 0
2022-12-13 10:30:46.880755+0100 vlobs[378:6102]    at: set_resource_name (drivers/vulkan/rendering_device_vulkan.cpp:8932)
2022-12-13 10:30:46.880923+0100 vlobs[378:6102] USER ERROR: All attachments unused.
2022-12-13 10:30:46.881003+0100 vlobs[378:6102]    at: framebuffer_create_multipass (drivers/vulkan/rendering_device_vulkan.cpp:4192)
```
Along with various other errors.

Basically I interpret this as failures to initialize various shaders and things for the renderer.
This then turns into errors like the following, which I interpret as basically things that are expected to now be in place are not.

```
2022-12-13 10:30:48.230373+0100 vlobs[378:6102] USER ERROR: Condition "!shader" is true. Returning: RID()
2022-12-13 10:30:48.230564+0100 vlobs[378:6102]    at: render_pipeline_create (drivers/vulkan/rendering_device_vulkan.cpp:6468)
2022-12-13 10:30:48.230650+0100 vlobs[378:6102] USER ERROR: Condition "pipeline.is_null()" is true. Returning: RID()
2022-12-13 10:30:48.230734+0100 vlobs[378:6102]    at: _generate_version (servers/rendering/renderer_rd/pipeline_cache_rd.cpp:61)
2022-12-13 10:30:48.230933+0100 vlobs[378:6102] USER ERROR: Condition "!pipeline" is true.
2022-12-13 10:30:48.231127+0100 vlobs[378:6102]    at: draw_list_bind_render_pipeline (drivers/vulkan/rendering_device_vulkan.cpp:7600)
2022-12-13 10:30:48.231304+0100 vlobs[378:6102] USER ERROR: Condition "!shader" is true. Returning: RID()
2022-12-13 10:30:48.231483+0100 vlobs[378:6102]    at: render_pipeline_create (drivers/vulkan/rendering_device_vulkan.cpp:6468)
2022-12-13 10:30:48.231715+0100 vlobs[378:6102] USER ERROR: Condition "pipeline.is_null()" is true. Returning: RID()
2022-12-13 10:30:48.231878+0100 vlobs[378:6102]    at: _generate_version (servers/rendering/renderer_rd/pipeline_cache_rd.cpp:61)
2022-12-13 10:30:48.232092+0100 vlobs[378:6102] USER ERROR: Condition "!pipeline" is true.
2022-12-13 10:30:48.232238+0100 vlobs[378:6102]    at: draw_list_bind_render_pipeline (drivers/vulkan/rendering_device_vulkan.cpp:7600)
```

And at this point the game crashes

It could well be that this hardware does not support running the engine with these features. So the issue here might more be that it is either not specifying and testing up front what features it requires (which would just block launching/installing the app) or correctly switching to fallbacks if features are missing and there are alternatives. Or, if this also fails, perhaps failing a bit more gracefully with an error message during startup.

### Steps to reproduce

This may be difficult to reproduce, since I think it might be specific to this hardware (and possibly other older iOS devices) and not reproducible without it. 

But I'm submitting this with this log in hopes that it might still be helpful as an example to see how failures during the engine initialization are not handled gracefully on iOS.

### Minimal reproduction project

See steps to reproduce

### PR DESCRIPTION
**Pull request by ChibiDenDen**
This changes simplifies the material_samplers arrays with constant elements and a single shared definition for all shaders.

Benefits:
reduces code duplication
removes sampler arrays in shaders (which are not supported by all devices)

_bugsquad edit: fixes: https://github.com/godotengine/godot/issues/70013_

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/samplers_inc.glsl
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: @RandomShaper will this play nice with D3D12? I'm a bit worried about setting bindings as high as 1000 on APIs that don't have the same set/binding model

Comment (ChibiDenDen) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: Is there a way for me to test the Dx12 path?

Comment (Calinou) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: @ChibiDenDen You can checkout https://github.com/godotengine/godot/pull/70315, apply your changes on top and follow the build instructions in the PR description.

Comment (ChibiDenDen) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: Ran the dx12 branch and cherry-picked my change, it seems to work fine (I get an error "Heap walker is at EOF" constantly with or without my change)
image

Comment (ChibiDenDen) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: Double checked with a texture mapped this time.
Dx12 is broken with this change.

Will update if I find a solution

Comment (ChibiDenDen) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: This is solved by either lowering GODOT_NIR_BINDING_MULTIPLIER

-static const uint32_t GODOT_NIR_BINDING_MULTIPLIER = 100000;
+static const uint32_t GODOT_NIR_BINDING_MULTIPLIER = 10000;

or lowering the binding numbers (lets say from 1000 to 100)

Screenshot from dx12 with texture:
image

Comment (RandomShaper) [reply servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: `100000` was a very conservative one. `10000` should be fine, too.

**Comment by ChibiDenDen** [comment]
> I'm quite concerned about this change. I'm not sure if it is safe to arbitrarily set a binding index so high (will we run into binding slot limits on some devices? Will we create a future headache for other APIs)). Further, I'm not even sure that this is solving a problem. There are no linked bug reports, so it looks like the old way of using a sampler array wasn't causing any issues.
> 
> You mention that sampler arrays aren't supported on all devices, could you elaborate on that? Have you tested on a device where master doesn't work, but this PR does? Alternatively, is this a limitation mentioned in the spec, or one we can look at in the GPUInfo database?

Hey @clayjohn 
To answer your questions

1.
according to the API documentation (https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkDescriptorSetLayoutBinding.html#_description)
The API allows using sparse binding numbers but claims that it "MAY" consume additional memory for the unused range.

Alternative options I thought of:
- Create a separate binding set for shared bindings (now it seems like only the samplers are shared by all shaders), but this will require updating the sets for all other bindings (which I have no issue doing if necessary).
- Create the sampler bindings dynamically in shader_compiler.cpp


2.
Im trying to run Godot4 vulkan on low end iPhone (iPhone 6 and iPhone simulator).
While GLES3 works, the performance on iOS is very bad compared to vulkan (which is based on metal), and awful on the simulator which uses software simulation for GLES3 and hardware acceleration for Metal.
Both have some limitations that can be seen here:
https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf

This changes resolves the limitation under "Array of samplers"

With this change and removing a single line in "vulkan_context.cpp"
```
	ERR_FAIL_COND_V_MSG(!physical_device_features.imageCubeArray, ERR_CANT_CREATE, "Your GPU doesn't support image cube arrays which are required to use the Vulkan-based renderers in Godot.");
```

Running 2D projects works on both devices with the vulkan renderer.

I would like to have even simple 3D projects working on the iPhone simulator if possible

3.
I did not open a bug about this but I would gladly open one if necessary.

4.
Regarding other devices which does not support "Array of samplers", I could not find any information, and I suspect that there are no issues with vulkan-native devices.

**Comment by ChibiDenDen** [comment]
The last update replaces the high binding index with a separate binding set.
The set number is dynamic and done similarly to the [render_forward_mobile.cpp](https://github.com/godotengine/godot/pull/77740/commits/6ea14cb0a081b4816d35691be0d27b06542fda35#diff-fb99395dba48cfb81f7e21da2a4f1e99fda0391e279402bf910ccee6a54eca11) MATERIAL_UNIFORM_SET define.

So at the cost of some additional management in every shader that requires the samplers
- adding the define
- maintaining an additional uniform set member variable
- binding the additional unifrom set

I removed the high index values.

I think that in general, using the defines to align the shader binding sets and the code values is a good practice.

**Comment by clayjohn** [comment]
That is a neat idea. Unfortunately, most android devices only support binding up to 4 sets at once https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxBoundDescriptorSets

This is why we have kept the number of sets so low in all of our shaders. It looks like iOS supports at least 8, so this approach would work there. But it will break on Android

**Comment by ChibiDenDen** [comment]
> That is a neat idea. Unfortunately, most android devices only support binding up to 4 sets at once https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxBoundDescriptorSets
> 
> This is why we have kept the number of sets so low in all of our shaders. It looks like iOS supports at least 8, so this approach would work there. But it will break on Android

Ohh wow, I did not know theres a limit on that.
Thanks for pointing this out.
I will rethink the approach here.

**Comment by ChibiDenDen** [comment]
Updated the code to not use an additional binding set, instead using a dynamic index assigned by each shader via a define.
Tested using multiple demo projects ( lights_and_shadows,  material_testers, particles, volumetric_fog) to cover the affected shaders, all works similarly with and without this change. (all tested on windows)

File: servers/rendering/renderer_rd/storage_rd/material_storage.cpp
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/material_storage.cpp:2859]:
```suggestion
	// Binding ids are aligned with samplers_inc.glsl.
```

**Review by clayjohn — APPROVED** [review]
This looks great now! I just left a small nitpick on one comment. but other than that, everything looks good. I am very happy with this approach now. 

I tested locally on a few test scenes I have and on GDQuests TPS demo and I didn't encounter any errors. 

This should be good to merge right after we release 4.1 so that it can get lots of testing before 4.2

File: servers/rendering/renderer_rd/shaders/samplers_inc.glsl
Comment (YuriSizov) [inline servers/rendering/renderer_rd/shaders/samplers_inc.glsl:1]: This file is missing an empty new line at the end.

**Comment by YuriSizov** [comment]
Oh, I merged this a bit too early by accident. It was still good, aside from a small formatting issue. But more importantly, I forgot to thank you for your contribution because of that. So thanks :)

**Pull request merged**
